tensorzero-optimizers/src/gepa/mod.rs

```diff
@@ -86,28 +89,275 @@ impl Optimizer for GEPAConfig {
         tracing::info!("Gateway client built successfully for GEPA optimization");
 
         // Initialize the Pareto frontier with baseline or provided variants
-        let pareto_frontier_variants = initialize_pareto_frontier(self, &function_context)?;
+        let initial_variants = initialize_pareto_frontier(self, &function_context)?;
 
         // Track original variant names to filter them out at the end
         let original_variant_names: std::collections::HashSet<String> =
-            pareto_frontier_variants.keys().cloned().collect();
+            initial_variants.keys().cloned().collect();
 
         tracing::info!(
             "Initialized with {} baseline variants: {:?}",
             original_variant_names.len(),
             original_variant_names
         );
 
+        // Create validation dataset for Pareto filtering
+        let val_dataset_name =
+            format!("{}_gepa_val_{}", self.evaluation_name, uuid::Uuid::now_v7());
+
+        // Track all temporary datasets for cleanup at the end
+        let mut temporary_datasets = vec![val_dataset_name.clone()];
+
+        tracing::info!(
+            "Creating validation dataset '{}' with {} examples",
+            val_dataset_name,
+            val_examples.len()
+        );
+
+        create_evaluation_dataset(
+            &config,
+            client,
+            clickhouse_connection_info,
+            val_examples,
+            &val_dataset_name,
+        )
+        .await?;
+
+        tracing::info!("Validation dataset created successfully");
+
+        // Evaluate initial variants on validation set
+        let num_variants = initial_variants.len();
+        tracing::info!(
+            "Evaluating {} initial variants on validation dataset",
+            num_variants
+        );
+
+        // Divide concurrency among variants to avoid max_concurrency² explosion
+        // Since each variant is evaluated on the same dataset, they'll take similar time
+        let per_variant_concurrency = (self.max_concurrency as usize / num_variants).max(1);
```

Comment:
really the right way to do this is pass around a semaphore. I'm not going to block this PR on that but please open an issue for it.